sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaIO.java

```diff
@@ -0,0 +1,307 @@
+     */
+    public Read withInputMetadata(Metadata metadata) {
+      Metadata inputMetadata = this.getInputMetadata();
+      if (inputMetadata != null) {
+        for (String name : metadata.names()) {
+            inputMetadata.set(name, metadata.get(name));
+        }
+      } else {
+        inputMetadata = metadata;
+      }
+      return toBuilder().setInputMetadata(inputMetadata).build();
+    }
+
+    /**
+     * Returns a new transform which will report the metadata.
+     */
+    public Read withReadOutputMetadata(Boolean value) {
+      return toBuilder().setReadOutputMetadata(value).build();
+    }
+
+    /**
+     * Returns a new transform which will use the specified queue poll time.
+     */
+    public Read withQueuePollTime(Long value) {
+      return toBuilder().setQueuePollTime(value).build();
+    }
+
+    /**
+     * Returns a new transform which will use the specified queue max poll time.
+     */
+    public Read withQueueMaxPollTime(Long value) {
+      return toBuilder().setQueueMaxPollTime(value).build();
+    }
+
+    /**
+     * Returns a new transform which will operate on the text blocks with the
+     * given minimum text length.
+     */
+    public Read withMinimumTextlength(Integer value) {
+      return toBuilder().setMinimumTextLength(value).build();
+    }
+
+    /**
+     * Returns a new transform which will use the synchronous reader.
+     */
+    public Read withParseSynchronously(Boolean value) {
+      return toBuilder().setParseSynchronously(value).build();
+    }
+
+    /**
+     * Path to Tika configuration resource.
+     */
+    public Read withOptions(TikaOptions options) {
+      checkNotNull(options, "TikaOptions cannot be empty.");
+      Builder builder = toBuilder();
+      builder.setFilepattern(StaticValueProvider.of(options.getInput()))
+             .setQueuePollTime(options.getQueuePollTime())
+             .setQueueMaxPollTime(options.getQueueMaxPollTime())
+             .setMinimumTextLength(options.getMinimumTextLength())
+             .setParseSynchronously(options.getParseSynchronously());
+      if (options.getContentTypeHint() != null) {
+        Metadata metadata = this.getInputMetadata();
+        if (metadata == null) {
+            metadata = new Metadata();
+        }
+        metadata.add(Metadata.CONTENT_TYPE, options.getContentTypeHint());
+        builder.setInputMetadata(metadata);
+      }
+      if (options.getTikaConfigPath() != null) {
+        builder.setTikaConfigPath(StaticValueProvider.of(options.getTikaConfigPath()));
+      }
+      if (Boolean.TRUE.equals(options.getReadOutputMetadata())) {
+        builder.setReadOutputMetadata(options.getReadOutputMetadata());
+      }
+      return builder.build();
+    }
+
+    @Override
+    public PCollection<String> expand(PBegin input) {
+      checkNotNull(this.getFilepattern(), "Filepattern cannot be empty.");
+      final Bounded<String> read = org.apache.beam.sdk.io.Read.from(new TikaSource(this));
```

Comment:
What were the design considerations behind choosing to implement this as a BoundedSource as opposed to a simple ParDo, especially on top of FileIO.readMatches() for example?

Reply:
Tika reports the data via the content handler, one by one, and I thought if we do it in a function then it would be a blocking point; TikaReader tries to avoid it